src/node/internal/internal_fs_utils.ts

```diff
@@ -0,0 +1,572 @@
+    throw new ERR_INVALID_ARG_TYPE(propName, 'ArrayBufferView[]', buffers);
+
+  for (let i = 0; i < buffers.length; i++) {
+    if (!isArrayBufferView(buffers[i]))
+      throw new ERR_INVALID_ARG_TYPE(propName, 'ArrayBufferView[]', buffers);
+  }
+
+  return buffers as ArrayBufferView[];
+}
+
+export function validateStringAfterArrayBufferView(
+  buffer: unknown,
+  name: string
+): void {
+  if (typeof buffer !== 'string') {
+    throw new ERR_INVALID_ARG_TYPE(
+      name,
+      ['string', 'Buffer', 'TypedArray', 'DataView'],
+      buffer
+    );
+  }
+}
+
+export function validateOffsetLengthWrite(
+  offset: number,
+  length: number,
+  byteLength: number
+): void {
+  if (offset > byteLength) {
+    throw new ERR_OUT_OF_RANGE('offset', `<= ${byteLength}`, offset);
+  }
+
+  if (length > byteLength - offset) {
+    throw new ERR_OUT_OF_RANGE('length', `<= ${byteLength - offset}`, length);
+  }
+
+  if (length < 0) {
+    throw new ERR_OUT_OF_RANGE('length', '>= 0', length);
+  }
+
+  validateInt32(length, 'length', 0);
+}
+
+// Our implementation of the Stats class differs a bit from Node.js' in that
+// the one in Node.js uses the older function-style class. However, use of
+// new fs.Stats(...) has been deprecated in Node.js for quite some time and
+// users really aren't supposed to be trying to create their own Stats objects.
+// Therefore, we intentionally use a class-style object here and make it an
+// error to try to create your own Stats object using the constructor.
+export class Stats {
+  public dev: number | bigint;
+  public ino: number | bigint;
+  public mode: number | bigint;
+  public nlink: number | bigint;
+  public uid: number | bigint;
+  public gid: number | bigint;
+  public rdev: number | bigint;
+  public size: number | bigint;
+  public blksize: number | bigint;
+  public blocks: number | bigint;
+  public atimeMs: number | bigint;
+  public mtimeMs: number | bigint;
+  public ctimeMs: number | bigint;
+  public birthtimeMs: number | bigint;
+  public atimeNs?: bigint;
+  public mtimeNs?: bigint;
+  public ctimeNs?: bigint;
+  public birthtimeNs?: bigint;
+  public atime: Date;
+  public mtime: Date;
+  public ctime: Date;
+  public birthtime: Date;
+
+  public constructor(
+    badge: symbol,
+    stat: InternalStat,
+    options: { bigint: boolean }
+  ) {
+    // The kBadge symbol is never exported for users. We use it as an internal
+    // marker to ensure that only internal code can create a Stats object using
+    // the constructor.
+    if (badge !== kBadge) {
+      throw new TypeError('Illegal constructor');
+    }
+
+    // All nodes are always readable
+    this.mode = 0o444;
+    if (stat.writable) {
+      this.mode |= 0o222; // writable
+    }
+
+    if (stat.device) {
+      this.mode |= S_IFCHR;
+    } else {
+      switch (stat.type) {
+        case 'file':
+          this.mode |= S_IFREG;
+          break;
+        case 'directory':
+          this.mode |= S_IFDIR;
```

Comment:
Should we also add execution permissions to directories here? Seems like that would usually be expected on a unix filesystem.

Reply:
We explicitly are not supporting execution permissions. If you call `access(...)` specifying X_OK it'll throw. If we ever do add any notion of execution permissions we can revisit this.